### PR TITLE
Support external partitions in pg_upgrade

### PR DESCRIPTION
**Pull request by danielgustafsson**
First some housekeeping notes for reviewing: this PR should be reviewed on a per-commit basis. I've tried to separate out the various bits of scaffolding required into separate commits to make it clearer what each part does. If there is a wish to break this down even further, either even smaller commits or multiple PRs, then please voice a concern.

Upgrading a cluster which has external tables as partitions, external partitions, is now supported whereas it previously made pg_upgrade error out and fail. The reason why external partitions require special handling is due to the fact that the partitioning system catalogs are not distributed to the segment servers. This means that partitioning is only supported as utility commands on the QD since it requires the catalogs. Since pg_upgrade operate on a per-segment basis in utility mode, this presents challenges. Further, external partitions cannot be defined as part of the table definition for partitioned tables, they instead need to be swapped into the partition hierarchy via `ALTER TABLE EXCHANGE` commands.

This patchset implements support for external partitions in pg_upgrade, and with it the required scaffolding to support it. I recommend starting the review by reading the included documentation in `contrib/pg_upgrade/README.gpdb`. The upgrade process can be broken down into the following pieces:

* The partitioning hierarchy is created intact by adding placeholder heap tables in lieu of the external partition, the external table is created separataly
* Oid synchronization for the objects are done by registering the new table (and associated types) as dispatch objects in pg_upgrade
* ALTER TABLE EXCHANGE commands are stored in a post-processing file as they cannot be applied during the upgrade

The scaffolding included in this patchset is:
* A post-processing Archive is added to pg_dump to store commands that must be executed after restore. This is for commands that are not allowed in utility mode.
* Allow pg_dump to create new objects as placeholders for existing objects in partitioning hierarchies.. Required on QE segments since the partitioning catalogs aren't distributed
* Refactoring of the existing arraytype Oid synchronization in pg_upgrade into a generic Oid synchronization dispatch

This PR depends on #1884 and #1885 to work for 4.3 to 5.0 upgrades, the commits for those PRs are not included in this PR however.

### CODE REVIEW

File: contrib/pg_upgrade/README.gpdb [hunk 1]
```diff
@@ -31,7 +40,7 @@ upgraded. This is because new Oids are allocated during the QD upgrade and
 these needs to be distributed to the QEs in order to preserve the QD/QE Oid
```
Comment (cjcjameson) [inline contrib/pg_upgrade/README.gpdb]: needs --> need (spelling)

File: contrib/pg_upgrade/README.gpdb [hunk 2]
```diff
@@ -31,7 +40,7 @@ upgraded. This is because new Oids are allocated during the QD upgrade and
...
 synchronization.
 
-* Each base relation type is in 5.0 defined as an array type, something which
+* Each base relation type is defined as an array type in 5.0, something which
   is not supported in 4.3. The arraytypes for base relations will be created
   during upgrade and thier Oids recorded. To be able to distribute the Oids to
```
Comment (cjcjameson) [inline contrib/pg_upgrade/README.gpdb]: their spelling

File: contrib/pg_upgrade/README.gpdb [hunk 3]
```diff
@@ -1,22 +1,31 @@
...
+The Greenplum Database pg_upgrade code is based on pg_upgrade from PostgreSQL
+9.0.23 and has been extended to work with Greenplum. This README is intended to
+document the changes that have been done and the Greenplum Database specific
+handling which is required to upgrade a Greenplum Database cluster.
 
-Operation of pg_upgrade remains unchanged as are the available command line
-parameters. Upgrading a Greenplum cluster with pg_upgrade is done by first
-upgrading the QD and then the QEs.
+The user-facing operation of pg_upgrade remains unchanged, as are the available
+command line parameters. The internals of pg_upgrade are however extended to
```
Comment (cjcjameson) [inline contrib/pg_upgrade/README.gpdb]: Is there anything specific to copy-mode vs. in-place mode that we need to call out?

Comment (danielgustafsson) [inline contrib/pg_upgrade/README.gpdb]: Nothing that comes to mind, this part works just as in upstream.

File: contrib/pg_upgrade/README.gpdb [hunk 4]
```diff
@@ -52,3 +61,168 @@ synchronization.
...
+
+      oid  |  relname  | relstorage
+    -------+-----------+------------
+     89387 | t         | h
+     89390 | t_x       | h
+     89407 | t_1_prt_2 | h
+     89423 | t_1_prt_3 | h
+     89450 | t_1_prt_1 | x
+
+At this point, t_x is a superfluous table and will in most cases be dropped.
```
Comment (danielgustafsson) [inline contrib/pg_upgrade/README.gpdb]: Operator error is the only case I can think of. It's easy to forget the DROP command when writing scripts, I did it myself many times over when working on this. Also, the syntax for exchanging a partition can probably look a bit scary, consider this example:
```SQL
CREATE EXTERNAL TABLE very_important_data .. ;
ALTER TABLE .. EXCHANGE .. WITH TABLE very_important_data;
DROP TABLE very_important_data;
```
I can imagine some tables left in there "just in case", no-one wants to be the guy who mistakenly dropped the important data.

File: src/bin/pg_dump/pg_dump.c [hunk 5]
```diff
@@ -9833,6 +9836,51 @@ dumpTableSchema(Archive *fout, TableInfo *tbinfo)
...
+		 * exist in the database we are dumping from, will be part of the
+		 * hierarchy during the restore process.
+		 *
+		 * Since TableInfo is a quite complicated structure, a deepcopy would
+		 * be not only complicated but all changes to the TableInfo would have
+		 * to be reflected. Since we need an exact copy of the external table
+		 * dobj but with another name, another relstorage and the partitioned
+		 * table as inheritance parent, temporarily scribble on the existing
+		 * dobj.
+		 */
```
Comment (danielgustafsson) [inline src/bin/pg_dump/pg_dump.c]: This is a hack.. and I would love ideas on how to improve this. Deepcopying something as involved as TableInfo turned ugly quick and will also require the deepcopy function to stay on equal footing with the struct as we merge upstream (which will undoubtedly be forgotten some time). Perhaps a shallow-copy with external points intact but with the altered values as an overlay?

**Comment by danielgustafsson** [comment]
Found a fixed a few bugs which are pushed as fixup commits, these will be squashed with their respective parent commits before going into master.

**Comment by Eulerizeit** [comment]
@slari-pivotal Can you review this PR today?

**Comment by danielgustafsson** [comment]
To test this PR I recommend doing the following:
1. Build and install this branch
2. Build and install a 4.3 branch
3. Create a gpdemo environment with `make cluster` in the 4.3 server, log in and create various tables, load/insert some data and then turn it off again
4. Create a gpdemo environment with `make cluster` in thr 5.0 server and turn it off immediately
5. Edit the `pg_hba.conf` files for both cluster to have `local   all   all   trust` line
6. Clone my upgrade script from: https://github.com/danielgustafsson/gpdemo_upgrade/blob/master/upgrade.sh
7. Edit the script to set your paths for the clusters and run `./upgrade.sh`

The 5.0 cluster should now have all the tables and data from the 4.3 cluster.

**Comment by hlinnaka** [comment]
The business of creating placeholder tables and exchanging them in sure is complicated..

I wonder if it would be simpler to not create the placeholder tables in the segments during upgrade, just leave them out. Instead, emit ALTER TABLE ADD PARTITION command through the master, just before the ALTER TABLE EXCHANGE PARTITION commands.

**Comment by hlinnaka** [comment]
The "Enforce catalog Oid in preassignment call" patch seems like a good idea regardless, +1 for pushing that separately and immediately.

**Comment by danielgustafsson** [comment]
> On 10 Mar 2017, at 14:14, Heikki Linnakangas <notifications@github.com> wrote:
> 
> The business of creating placeholder tables and exchanging them in sure is complicated..

I’m not going to argue with that =)

> I wonder if it would be simpler to not create the placeholder tables in the segments during upgrade, just leave them out. Instead, emit ALTER TABLE ADD PARTITION command through the master, just before the ALTER TABLE EXCHANGE PARTITION commands.

I tried that first, but generating the correct PARTITION statement such that the external partitions could be added in into the right place (to make the restored table 100% identical to the dumped) proved even more complicated.

**Comment by hlinnaka** [comment]
On 03/10/2017 03:27 PM, Daniel Gustafsson wrote:
>> On 10 Mar 2017, at 14:14, Heikki Linnakangas
>> <notifications@github.com> wrote: I wonder if it would be simpler
>> to not create the placeholder tables in the segments during
>> upgrade, just leave them out. Instead, emit ALTER TABLE ADD
>> PARTITION command through the master, just before the ALTER TABLE
>> EXCHANGE PARTITION commands.
>
> I tried that first, but generating the correct PARTITION statement
> such that the external partitions could be added in into the right
> place (to make the restored table 100% identical to the dumped)
> proved even more complicated.

Ok.

- Heikki

File: contrib/pg_upgrade/README.gpdb [hunk 6]
```diff
@@ -10,13 +10,19 @@ upgrading the QD and then the QEs.
...
 ---------------------------------
 
-pg_upgrade is in PostgreSQL only maintaning the Oids of types, Greenplum
-however requires all Oids to be maintained such that they are synchronized
-between QD and QEs after upgrade. The binary-upgrade option in pg_dump has
-been extended to inject Oid preservation calls for all required object types
-in the dump. Technically, the Oids needn't be preserved during upgrade as
-long as they are synchronized between QD/QEs after the upgrade. The current
-implementation preserves the Oids rather than allocate new ones.
+pg_upgrade is in PostgreSQL only maintaining the Oids of types across an
```
Comment (hlinnaka) [inline contrib/pg_upgrade/README.gpdb]: would reword as "In PostgreSQL, pg_upgrade only maintains ..."

File: contrib/pg_upgrade/dump.c [hunk 7]
```diff
@@ -9,6 +9,77 @@
...
+			for (i = 0; i < active_db->dispatch_arr.ndispatch; i++)
+			{
+				DispatchInfo	*d = &active_db->dispatch_arr.dispatches[i];
+
+				if (d->type == DISPATCH_ARRAYTYPE)
+				{
+					fprintf(script,
+							"SELECT binary_upgrade.preassign_arraytype_oid("
+								"'%u'::pg_catalog.oid, '%s'::text, '%u'::pg_catalog.oid);\n",
+							 d->oid, d->name, d->namespace_oid);
```
Comment (hlinnaka) [inline contrib/pg_upgrade/dump.c]: does d->name need to be quoted/escaped here?

File: contrib/pg_upgrade/check.c [hunk 8]
```diff
@@ -769,3 +769,142 @@ check_for_reg_data_type_usage(migratorContext *ctx, Cluster whichCluster)
 	else
 		check_ok(ctx);
 }
+
+/*
+ *	exclude_oid_for_dispatch
+ *
+ *	Generate a string suitable for the relation query in get_rel_info() to
+ *	exclude any relation stored as a dispatch object. The reason for excluding
+ *	them is that any table object registered as an Oid dispatch will be
```
Comment (hlinnaka) [inline contrib/pg_upgrade/check.c]: s/be definition/by definition/

Comment (hlinnaka) [inline contrib/pg_upgrade/check.c]: Suggest renaming this to something like exclude_rel_oids_for_dispatch(), to make it clear that this is about relations, not any other objects.

File: contrib/pg_upgrade/check.c [hunk 9]
```diff
@@ -769,3 +769,142 @@ check_for_reg_data_type_usage(migratorContext *ctx, Cluster whichCluster)
...
+		for (dispno = 0; dispno < db->dispatch_arr.ndispatch; dispno++)
+		{
+			DispatchInfo   *rel = &db->dispatch_arr.dispatches[dispno];
+			int				ret;
+
+			if (rel->type == DISPATCH_RELATION)
+			{
+				ret = snprintf(p, sizeof(query) - (p - query), " AND relname <> '%s'::text ", rel->name);
+				if (ret >= (sizeof(query) - (p - query)))
+					pg_log(ctx, PG_FATAL, "Dispatch exclusion too large");
```
Comment (hlinnaka) [inline contrib/pg_upgrade/check.c]: It doesn't seem impossible to really run out of space here, if you have a huge schema.

**Comment by hlinnaka** [comment]
I think it would make sense to submit a separate PR for the "Refactor arraytype dispatch into generic Oid dispatch" patch. That's already quite big, and seems like a reasonable thing even without the rest of the PR (although we probably wouldn't bother, if we didn't need the generality for other things),

**Comment by hlinnaka** [comment]
So, if I understand this correctly, in both the segments and the master, we create a separate external table for any external partition, and a placeholder heap table. And once the whole upgrade on the cluster is done, we run ALTER TABLE EXCHANGE PARTITION commands to swap the external tables with the heap partitions.

I'm trying to figure out if there really is no simpler way. We basically need to restore the state of the catalogs in the segment to what it was in the old cluster, while in utility mode. We can CREATE EXTERNAL TABLE in utility mode, but we cannot ALTER TABLE EXCHANGE PARTITION, because partition rules are not stored in segments. But that also means that we don't need to restore partition rules in segments.

(Fun fact: you can run `CREATE TABLE ... PARTITION BY ..` in a single segment in utility mode. And if you do that, rows will appear in pg_partition_rule.)

How do we dump partitions in a segment, currently? As `CREATE TABLE <partition> ... INHERITS <parent>`? Why can't we do that for an external table? CREATE EXTERNAL TABLE doesn't support INHERITS, at the moment, but would it be easy to add that, for binary upgrade?

Or, would it be easy to emit suitable INSERT/UPDATE commands to hack the catalogs directly, for the state we want?

**Comment by hlinnaka** [comment]
> How do we dump partitions in a segment, currently? As CREATE TABLE <partition> ... INHERITS <parent>? Why can't we do that for an external table? CREATE EXTERNAL TABLE doesn't support INHERITS, at the moment, but would it be easy to add that, for binary upgrade?

Or perhaps it would be simpler to emit `CREATE EXTERNAL TABLE`, followed by `ALTER TABLE exttabe INHERITS ...`. I just quickly tested that, and the only thing that stops that from working is that we explicitly forbid that. It would be a one-line change to allow it to work in binary-upgrade mode or utility mode.

**Comment by danielgustafsson** [comment]
> On 10 Mar 2017, at 15:51, Heikki Linnakangas <notifications@github.com> wrote:
> 
> So, if I understand this correctly, in both the segments and the master, we create a separate external table for any external partition, and a placeholder heap table. And once the whole upgrade on the cluster is done, we run ALTER TABLE EXCHANGE PARTITION commands to swap the external tables with the heap partitions.

Correct.

> I'm trying to figure out if there really is no simpler way. We basically need to restore the state of the catalogs in the segment to what it was in the old cluster, while in utility mode. We can CREATE EXTERNAL TABLE in utility mode, but we cannot ALTER TABLE EXCHANGE PARTITION, because partition rules are not stored in segments. But that also means that we don't need to restore partition rules in segments.

This patch performs the upgrade the sort of “dumb” way by automating lots of steps that you could do manually essentially. In part, this is by design to fully understand the problem domain (which was more tangled than I thought). If we can figure out a better way now that the process is laid out, I’m all for pursuing that and removing code from here.

> (Fun fact: you can run CREATE TABLE ... PARTITION BY .. in a single segment in utility mode. And if you do that, rows will appear in pg_partition_rule.)

That sounds like a bug..

> How do we dump partitions in a segment, currently? As CREATE TABLE <partition> ... INHERITS <parent>? Why can't we do that for an external table? CREATE EXTERNAL TABLE doesn't support INHERITS, at the moment, but would it be easy to add that, for binary upgrade?

For the master we still can’t do a CREATE TABLE .. PARTITON BY that includes the external tables as external partitions aren’t supported in partitioning definitions. So we still need the ALTER TABLE .. EXCHANGE, or am I missing something?

> Or, would it be easy to emit suitable INSERT/UPDATE commands to hack the catalogs directly, for the state we want?

Thats another option, and the reason for not doing it (yet) essentially boils down to the above statement on not wanting to be too clever before understanding what to be clever about =) Question is also how long we’d need this given the upstream merges that are pending, perhaps it’s only for a single release?

**Comment by danielgustafsson** [comment]
> On 10 Mar 2017, at 14:59, Heikki Linnakangas <notifications@github.com> wrote:
> 
> @hlinnaka commented on this pull request.
> 
> In contrib/pg_upgrade/check.c:
> 
> > +		if (strcmp(db->db_name, master_db->db_name) == 0)
> +			break;
> +	}
> +
> +	if (db->dispatch_arr.ndispatch > 0)
> +	{
> +		for (dispno = 0; dispno < db->dispatch_arr.ndispatch; dispno++)
> +		{
> +			DispatchInfo   *rel = &db->dispatch_arr.dispatches[dispno];
> +			int				ret;
> +
> +			if (rel->type == DISPATCH_RELATION)
> +			{
> +				ret = snprintf(p, sizeof(query) - (p - query), " AND relname <> '%s'::text ", rel->name);
> +				if (ret >= (sizeof(query) - (p - query)))
> +					pg_log(ctx, PG_FATAL, "Dispatch exclusion too large");
> 
> It doesn't seem impossible to really run out of space here, if you have a huge schema.

Not at all, and it indeed needs to be revisited. I thought I had left a comment highlighting that there but apparently I had missed that =/

**Comment by Eulerizeit** [comment]
> (Fun fact: you can run CREATE TABLE ... PARTITION BY .. in a single segment in utility mode. And if you do that, rows will appear in pg_partition_rule.)

This seems like a bad thing.

**Comment by hlinnaka** [comment]
On 03/10/2017 05:14 PM, Robert Eckhardt wrote:
>> (Fun fact: you can run CREATE TABLE ... PARTITION BY .. in a single segment in utility mode. And if you do that, rows will appear in pg_partition_rule.)
>
> This seems like a bad thing.

Well, there's a lot of ways to shoot yourself in the foot in utility 
mode. But yeah, perhaps we should forbid that particular footgun, I 
don't see any valid use case for that.

- Heikki

**Comment by Eulerizeit** [comment]
We are going to be revisiting, maybe, I hope, declarative partitioning to
at least evaluate how we will deal with the upstream changes.  Since that
is a 10 thing I don't feel a lot of pressure but I'll add this tidbit
there.

**Comment by hlinnaka** [comment]
On 03/10/2017 05:08 PM, Daniel Gustafsson wrote:
>> On 10 Mar 2017, at 15:51, Heikki Linnakangas  wrote:
>>
>> How do we dump partitions in a segment, currently? As CREATE TABLE
>> <partition> ... INHERITS <parent>? Why can't we do that for an
>> external table? CREATE EXTERNAL TABLE doesn't support INHERITS, at
>> the moment, but would it be easy to add that, for binary upgrade?
>
> For the master we still can’t do a CREATE TABLE .. PARTITON BY that
> includes the external tables as external partitions aren’t supported
> in partitioning definitions. So we still need the ALTER TABLE ..
> EXCHANGE, or am I missing something?

Hmm, yeah. Would it still be simpler to do that in the segments, and 
only use EXCHANGE in the master? You wouldn't need to synchronize the 
OIDs of the placeholder heaps.

>> Or, would it be easy to emit suitable INSERT/UPDATE commands to
>> hack the catalogs directly, for the state we want?
>
> Thats another option, and the reason for not doing it (yet)
> essentially boils down to the above statement on not wanting to be
> too clever before understanding what to be clever about =)

Agreed.

> Question is also how long we’d need this given the upstream merges
> that are pending, perhaps it’s only for a single release?

Yeah, going from GPDB 5 to GPDB 6, all the external table commants will 
hopefully be translated into corresponding foreign data wrapper and 
foreign table commands. That'll be even more fun :-)

- Heikki

**Comment by danielgustafsson** [comment]
> On 10 Mar 2017, at 16:28, Heikki Linnakangas <notifications@github.com> wrote:
> 
> On 03/10/2017 05:08 PM, Daniel Gustafsson wrote:
> >> On 10 Mar 2017, at 15:51, Heikki Linnakangas wrote:
> >>
> >> How do we dump partitions in a segment, currently? As CREATE TABLE
> >> <partition> ... INHERITS <parent>? Why can't we do that for an
> >> external table? CREATE EXTERNAL TABLE doesn't support INHERITS, at
> >> the moment, but would it be easy to add that, for binary upgrade?
> >
> > For the master we still can’t do a CREATE TABLE .. PARTITON BY that
> > includes the external tables as external partitions aren’t supported
> > in partitioning definitions. So we still need the ALTER TABLE ..
> > EXCHANGE, or am I missing something?
> 
> Hmm, yeah. Would it still be simpler to do that in the segments, and 
> only use EXCHANGE in the master? You wouldn't need to synchronize the 
> OIDs of the placeholder heaps.

Perhaps. We’d need to make EXCHANGE allowed in utility mode as well which it currently isn’t, and DROP the superflous table (ao/heap) which was exchanged out of the partitioning, to ensure a consistent catalog across QD/QEs. It might make it easier if we always make the placeholder relation heap, regardless of the original partitioning clause, then we at least only have a single tabletype to handle?

**Comment by danielgustafsson** [comment]
> On 11 Mar 2017, at 10:37, Daniel Gustafsson <daniel@yesql.se> wrote:
> 
> 
>> On 10 Mar 2017, at 16:28, Heikki Linnakangas <notifications@github.com> wrote:
>> 
>> On 03/10/2017 05:08 PM, Daniel Gustafsson wrote:
>>>> On 10 Mar 2017, at 15:51, Heikki Linnakangas wrote:
>>>> 
>>>> How do we dump partitions in a segment, currently? As CREATE TABLE
>>>> <partition> ... INHERITS <parent>? Why can't we do that for an
>>>> external table? CREATE EXTERNAL TABLE doesn't support INHERITS, at
>>>> the moment, but would it be easy to add that, for binary upgrade?
>>> 
>>> For the master we still can’t do a CREATE TABLE .. PARTITON BY that
>>> includes the external tables as external partitions aren’t supported
>>> in partitioning definitions. So we still need the ALTER TABLE ..
>>> EXCHANGE, or am I missing something?
>> 
>> Hmm, yeah. Would it still be simpler to do that in the segments, and 
>> only use EXCHANGE in the master? You wouldn't need to synchronize the 
>> OIDs of the placeholder heaps.
> 
> Perhaps. We’d need to make EXCHANGE allowed in utility mode as well which it currently isn’t, and DROP the superflous table (ao/heap) which was exchanged out of the partitioning, to ensure a consistent catalog across QD/QEs. It might make it easier if we always make the placeholder relation heap, regardless of the original partitioning clause, then we at least only have a single tabletype to handle?

On that note, would we in the QEs need to “burn” the Oids consumed in the QD for the placeholder? It seems safe to skip that given how Oid assignment works but I haven’t done the research yet.

**Comment by danielgustafsson** [comment]
> On 10 Mar 2017, at 14:23, Heikki Linnakangas <notifications@github.com> wrote:
> 
> The "Enforce catalog Oid in preassignment call" patch seems like a good idea regardless, +1 for pushing that separately and immediately.

Thanks, pushed that to master and will rebase away from this PR with the next round of fixes.

**Comment by Eulerizeit** [comment]
@danielgustafsson let's close this.

**Comment by danielgustafsson** [comment]
Thanks for the reminder, had forgotten to do that. Closing as it's replaced by #2053.
